hcloud/instances.go

```diff
@@ -20,48 +20,99 @@ import (
+	"github.com/hetznercloud/hcloud-cloud-controller-manager/internal/robot"
 	"github.com/hetznercloud/hcloud-go/v2/hcloud"
 )
 
 type instances struct {
 	client        *hcloud.Client
+	robotClient   robot.Client
 	addressFamily config.AddressFamily
 	networkID     int64
 }
 
-func newInstances(client *hcloud.Client, addressFamily config.AddressFamily, networkID int64) *instances {
-	return &instances{client, addressFamily, networkID}
+var errServerNotFound = fmt.Errorf("server not found")
+
+func newInstances(client *hcloud.Client, robotClient robot.Client, addressFamily config.AddressFamily, networkID int64) *instances {
+	return &instances{client, robotClient, addressFamily, networkID}
 }
 
-// lookupServer attempts to locate the corresponding hcloud.Server for a given corev1.Node
+// lookupServer attempts to locate the corresponding [*hcloud.Server] or [*robotmodels.Server] for a given [*corev1.Node].
 // It returns an error if the Node has an invalid provider ID or if API requests failed.
-// It can return a nil [*hcloud.Server] if neither the ProviderID nor the Name matches an existing server.
-func (i *instances) lookupServer(ctx context.Context, node *corev1.Node) (*hcloud.Server, error) {
-	var server *hcloud.Server
+// It can return nil server if neither the ProviderID nor the Name matches an existing server.
+func (i *instances) lookupServer(
+	ctx context.Context,
+	node *corev1.Node,
+) (genericServer, error) {
 	if node.Spec.ProviderID != "" {
-		serverID, err := providerid.ToServerID(node.Spec.ProviderID)
+		var serverID int64
+		serverID, isCloudServer, err := providerid.ToServerID(node.Spec.ProviderID)
+
 		if err != nil {
 			return nil, fmt.Errorf("failed to convert provider id to server id: %w", err)
 		}
 
-		server, _, err = i.client.Server.GetByID(ctx, serverID)
+		if isCloudServer {
+			server, err := getCloudServerByID(ctx, i.client, serverID)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get hcloud server \"%d\": %w", serverID, err)
+			}
+
+			if server == nil {
+				return nil, nil
+			}
+
+			return hcloudServer{server}, nil
+		}
+
+		if i.robotClient == nil {
+			return nil, errMissingRobotCredentials
+		}
+		server, err := getRobotServerByID(i.robotClient, int(serverID), node)
 		if err != nil {
-			return nil, fmt.Errorf("failed to lookup server \"%d\": %w", serverID, err)
+			return nil, fmt.Errorf("failed to get robot server \"%d\": %w", serverID, err)
 		}
-	} else {
-		var err error
-		server, _, err = i.client.Server.GetByName(ctx, node.Name)
+
+		if server == nil {
+			return nil, nil
+		}
+
+		return robotServer{server, i.robotClient}, nil
+	}
+
+	// If the node has no provider ID we try to find the server by name from
+	// both sources. In case we find two servers, we return an error.
+	cloudServer, err := getCloudServerByName(ctx, i.client, node.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get hcloud server %q: %w", node.Name, err)
+	}
+
+	var rServer *robotmodels.Server
```

Comment:
```suggestion
	var robotServer *robotmodels.Server
```

Reply:
The identifier `robotServer` is already used by the struct implementing `genericServer`.

Reply:
Maybe use the `h` to determine if it's a local struct or a struct from the hrobot library ?